text/0225-configuration.md

```diff
@@ -0,0 +1,411 @@
+          #
+          # Environment variable: OTEL_TRACES_EXPORTER
+          exporter: zipkin
+  # Configure the meter provider.
+  ...
+```
+
+### Environment variable substitution
+
+Configuration files *MUST* support environment variable expansion. While this accommodates the scenario in which a configuration file needs to reference sensitive data and is not able to be stored securely, environment variable expansion is not limited to sensitive data.
+
+The syntax for environment variable expansion mirrors the [collector](https://opentelemetry.io/docs/collector/configuration/#configuration-environment-variables).
+
+For example, given an environment where `API_KEY=1234`, the configuration file contents:
+
+```yaml
+scheme_version: 0.1
+sdk:
+  tracer_provider:
+    exporters:
+      otlp:
+        endpoint: https://example.host:4317/v1/traces
+        headers:
+          api-key: ${env:API_KEY}
+```
+
+Result in the following after substitution:
+
+```yaml
+scheme_version: 0.1
+sdk:
+  tracer_provider:
+    exporters:
+      otlp:
+        endpoint: https://example.host:4317/v1/traces
+        headers:
+          api-key: 1234
+```
+
+Implementations *MUST* perform environment variable substitution before validating and parsing configuration file contents.
+
+### Version guarantees & backwards compatibility
+
+Each version of the configuration schema carries a major and minor version. Configurations specify the major and minor version they adhere to. Before reaching 1.0, each minor version change is equivalent to major version change. That is, there are no guarantees about compatibility and all changes are permitted. As of 1.0, we provide the following stability guarantees:
+
+* For major version: No guarantees.
+* For minor versions:
+  * Property keys will not change. Although the addition of optional properties is permitted and the removal of properties is permitted, this *MUST* not be abused to rename property keys.
+  * Property value types will not change, except that integers MAY become floating points.
+  * No additional required properties.
+
+Allowable changes:
+
+* For major versions: All changes are permitted.
+* For minor versions:
+  * Addition of optional properties.
+  * Required property keys may become optional.
+  * Removal of properties, provided that the property key is not reused in the future.
+
+SDKs validating configuration *MUST* fail when they encounter a configuration with an unsupported version. Generally, this means fail when encountering a major version which is not recognized. An SDK might choose to maintain a library of validators / parsers for each major version, and use the configuration version to select and use the correct instance. Differences in minor versions (except pre-1.0 minor versions) SHOULD be acceptable, with the caveat that allowable property additions and removals MAY result in configuration that is different than excepted.
```

Comment:
> Differences in minor versions (except pre-1.0 minor versions) SHOULD be acceptable, with the caveat that allowable property additions and removals MAY result in configuration that is different than excepted.

Should this be `MUST be acceptable`?  If I have a config with the same major version as a parser (but different minor/patch), those MUST work together by definition, as no breaking changes should have occurred.

Or is this saying that using a config that is newer than the parser will result in the parser dropping fields it doesn't know about?

Reply:
> Or is this saying that using a config that is newer than the parser will result in the parser dropping fields it doesn't know about?

Yes, that was intent of the sentence.

Reply:
I agree that this *SHOULD* should be a *MUST* since parser and config versions of the same major version must work together.